### PR TITLE
add unit test for traversers api

### PR DESCRIPTION
**Pull request by zyxxoo**
add unit test (allshortestPath, countAPI, CrosspointsAPI, CustomizedCrosspointsAPI)

### CODE REVIEW

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/AllShortestPathsAPITest.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Map;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableMap;
+
+
+public class AllShortestPathsAPITest extends BaseApiTest {
+    public static String path = "graphs/hugegraph/traversers/allshortestpaths";
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/AllShortestPathsAPITest.java:12]: fix code style https://github.com/hugegraph/hugegraph-doc/wiki/HugeGraph%E4%BB%A3%E7%A0%81%E9%A3%8E%E6%A0%BC%E6%8C%87%E5%8D%97

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 2]
```diff
@@ -127,11 +133,41 @@ public Response get(String path, String id) {
         }
 
         public Response get(String path, Map<String, Object> params) {
-            WebTarget target = this.target.path(path);
-            for (Map.Entry<String, Object> i : params.entrySet()) {
-                target = target.queryParam(i.getKey(), i.getValue());
+            return this.get(path,
+                            params.entrySet().stream()
+                                  .map(entry -> new Pair<String, Object>(
+                                          entry.getKey(), entry.getValue()))
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:139]: improve align

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 3]
```diff
@@ -284,6 +320,71 @@ protected static void initIndexLabel() {
         assertResponseStatus(202, r);
     }
 
+    protected static void initEdge() {
+        String path = URL_PREFIX + GRAPH_EDGE + BATCH;
+        Map<String, String> ret = getAllName2VertexIds();
+        System.out.println(ret);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:326]: not allowed println

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 4]
```diff
@@ -349,8 +450,28 @@ protected static Response createAndAssert(String path, String body) {
...
+        List<Map> list = readList(content, "vertices", Map.class);
+
+        return list.stream().filter(map -> map.get("properties") != null &&
+                                           ((Map) map.get("properties"))
+                                                   .get("name") != null)
+                   .collect(
+                           Collectors.toMap(map -> ((Map) map.get("properties"))
+                                                    .get(
+                                                            "name").toString(),
+                                            map -> map.get("id").toString()));
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:470]: align

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 5]
```diff
@@ -349,8 +450,28 @@ protected static Response createAndAssert(String path, String body) {
         return r;
     }
 
+    protected static Map<String, String> getAllName2VertexIds() {
+        Response r = client.get(URL_PREFIX + GRAPH_VERTEX);
+        String content = r.readEntity(String.class);
+        if (r.getStatus() != 200) {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:456]: assert them

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        BaseApiTest.initEdge();
+    }
+
+    @Test
+    public void testGet() {
+        Map<String, String> name2Ids = getAllName2VertexIds();
+        String markoId = name2Ids.get("marko");
+        String vadasId=  name2Ids.get("vadas");
+        Map<String, Object> map = ImmutableMap.of("source",
+                                                  String.format("\"%s\"",
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java:31]: add method id2Json()

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 7]
```diff
@@ -349,18 +455,34 @@ protected static Response createAndAssert(String path, String body) {
         return r;
     }
 
+    protected static Map<String, String> getAllName2VertexIds() {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:458]: allVertexName2IdMap

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        BaseApiTest.initVertex();
+        BaseApiTest.initEdge();
+    }
+
+    @Test
+    public void testGet() {
+        Map<String, String> name2Ids = getAllName2VertexIds();
+        String markoId = name2Ids.get("marko");
+        String vadasId=  name2Ids.get("vadas");
+        Map<String, Object> map = ImmutableMap.of("source",
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java:30]: rename to params

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        BaseApiTest.initEdgeLabel();
+        BaseApiTest.initVertex();
+        BaseApiTest.initEdge();
+    }
+
+    @Test
+    public void testCount() {
+        String markoId = getAllName2VertexIds().get("marko");
+        String content = String.format("{\n" +
+                                       "\t\"source\": \"%s\",\n" +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java:31]: improve style

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CrosspointsAPITest.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                                                                   markoId),
+                                                     "target",
+                                                     String.format("\"%s\"",
+                                                                   vadasId),
+                                                     "max_depth", 1000);
+        Response r = client().get(path, params);
+        assertResponseStatus(200, r);
+        Map<String, Object> data = r.readEntity(Map.class);
+        List<Map<String, Object>> crosspoints =
+                (List<Map<String, Object>>) data.get("crosspoints");
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CrosspointsAPITest.java:44]: call assertContains

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 11]
```diff
@@ -450,24 +427,36 @@ protected static Response createAndAssert(String path, String body) {
...
-        List<Map> list = readList(content, "vertices", Map.class);
+        List<Map> vertices = readList(content, "vertices", Map.class);
+
+        Map<String, String> ret = new HashMap<>();
+        for (Map vertice : vertices) {
+            Map properties = (Map) vertice.get("properties");
+            if (properties == null ||
+                !properties.containsKey("name") ||
+                !vertice.containsKey("id")) {
+
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:443]: empty line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 12]
```diff
@@ -450,24 +427,36 @@ protected static Response createAndAssert(String path, String body) {
         return r;
     }
 
-    protected static Map<String, String> getAllName2VertexIds() {
+    protected static Map<String, String> listAllName2VertexIds() {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:430]: listAllVertexName2Ids

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 13]
```diff
@@ -450,24 +427,36 @@ protected static Response createAndAssert(String path, String body) {
...
-        if (r.getStatus() != 200) {
-            throw new HugeException("Failed to get vertex id: %s", content);
-        }
+        String content = assertResponseStatus(200, r);
 
         @SuppressWarnings("rawtypes")
-        List<Map> list = readList(content, "vertices", Map.class);
+        List<Map> vertices = readList(content, "vertices", Map.class);
+
+        Map<String, String> ret = new HashMap<>();
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:437]: improve var name

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java (hunk 6 above)
Comment (javeme) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java:31]: not addressed: add method id2Json()

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+                                       "  }, " +
+                                       "   { " +
+                                       "   \"labels\": [ " +
+                                       "    ], " +
+                                       "    \"degree\": 100, " +
+                                       "    \"skip_degree\": 100 " +
+                                       "  }  " +
+                                       "   " +
+                                       " ] " +
+                                       "}", markoId);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java:53]: keep the same style

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/FusiformSimilarityAPITest.java [hunk 15]
```diff
@@ -24,31 +25,31 @@ public void prepareSchema() {
...
-                                         "    \"with_intermediary\": false,\n" +
-                                         "    \"with_vertex\":true\n" +
+    public void testPost() {
+        Response r = client().post(path, "{ " +
+                                         "    \"sources\":{ " +
+                                         "        \"ids\":[], " +
+                                         "        \"label\": \"person\", " +
+                                         "        \"properties\": { " +
+                                         "        } " +
+                                         "    }, " +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/FusiformSimilarityAPITest.java:35]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesAPITest.java [hunk 16]
```diff
@@ -33,35 +34,35 @@ public void prepareSchema() {
...
         final String edgeGetPath = "graphs/hugegraph/graph/edges";
         String vadasId = name2Ids.get("vadas");
-        Response r = client().get(edgeGetPath, ImmutableMap.of("vertex_id",
-                                                               "\"" + vadasId + "\"",
-                                                               "direction",
-                                                               "IN"));
-        assertEquals(200, r.getStatus());
-        Map entity = r.readEntity(Map.class);
+        Response r = client().get(edgeGetPath,
+                                  ImmutableMap.of("vertex_id",
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesAPITest.java:42]: define var params first

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CustomizedCrosspointsAPITest.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+                                       "        { " +
+                                       "            \"steps\":[ " +
+                                       "                { " +
+                                       "                    \"direction\":\"BOTH\"," +
+                                       "                    \"labels\":[ " +
+                                       "                    ], " +
+                                       "                    \"degree\":-1 " +
+                                       "                } " +
+                                       "            ] " +
+                                       "        } " +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CustomizedCrosspointsAPITest.java:49]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/ApiTestSuite.java [hunk 18]
```diff
@@ -23,10 +23,25 @@
...
+import com.baidu.hugegraph.api.traversers.MultiNodeShortestPathAPITest;
+import com.baidu.hugegraph.api.traversers.NeighborRankAPITest;
+import com.baidu.hugegraph.api.traversers.PathsAPITest;
+import com.baidu.hugegraph.api.traversers.PersonalRankAPITest;
+import com.baidu.hugegraph.api.traversers.RaysAPITest;
 import com.baidu.hugegraph.dist.RegisterUtil;
 
 @RunWith(Suite.class)
 @Suite.SuiteClasses({
+    UserApiTest.class,
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/ApiTestSuite.java:44]: add TraversersAPITestSuite

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/TraversersAPITestSuite.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    JaccardSimilarityAPITest.class,
+    KneighborAPITest.class,
+    KoutAPITest.class,
+    MultiNodeShortestPathAPITest.class,
+    NeighborRankAPITest.class,
+    PathsAPITest.class,
+    PersonalRankAPITest.class,
+    RaysAPITest.class
+})
+public class TraversersAPITestSuite {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/TraversersAPITestSuite.java:26]: rename to TraversersApiTestSuite

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/JaccardSimilarityAPITest.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Test
+    public void testGet() {
+        Map<String, String> name2Ids = listAllName2VertexIds();
+        String markoId = name2Ids.get("marko");
+        String peterId = name2Ids.get("peter");
+        Response r = client().get(path, ImmutableMap.of("vertex",
+                                                        String.format("\"%s\"",
+                                                                      markoId),
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/JaccardSimilarityAPITest.java:36]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 21]
```diff
@@ -603,6 +590,10 @@ protected static String assertResponseStatus(int status,
         return content;
     }
 
+    protected static Map<String, Object> parseMap(String json) {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:593]: delete this method and call assertJsonContains() instead

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        Response r = client().post(path, reqBody);
+
+        String content = assertResponseStatus(200, r);
+        Map<String, Object> entity = parseMap(content);
+        Integer count = (Integer) entity.get("count");
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java:58]: call assertJsonContains(content, "count")

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java (hunk 6 above)
Comment (zyxxoo) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java:31]: 之前有考虑过封装， 但是想这里看起来直观一点就么封装函数了

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java (hunk 21 above)
Comment (zyxxoo) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:589]: 这里可以考虑下保留，assertJsonContains 适合只有一个key的地方，多个key的时候 assertJson 要解析多次，而且有的时候只需要验证size >0 就可以，用 assertJsonContains 做不到，我现在将只需拿一个值的地方都改成了 assertJsonContains, 其他地方还是用的 解析 map的方式来做

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 23]
```diff
@@ -349,18 +425,50 @@ protected static Response createAndAssert(String path, String body) {
...
+        return vertextName2Ids;
+    }
+
+    protected static String id2Json(String params) {
+        return String.format("\"%s\"", params);
+    }
+
     protected static String getVertexId(String label, String key, String value)
-                                        throws IOException {
+            throws IOException {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:464]: align

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 24]
```diff
@@ -448,14 +556,14 @@ protected static String parseId(String content) throws IOException {
             JsonNode element = root.get(key);
             if (element == null) {
                 throw new HugeException(String.format(
-                          "Can't find value of the key: %s in json.", key));
+                        "Can't find value of the key: %s in json.", key));
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:559]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 25]
```diff
@@ -448,14 +556,14 @@ protected static String parseId(String content) throws IOException {
...
             }
             JavaType type = mapper.getTypeFactory()
                                   .constructParametricType(List.class, clazz);
             return mapper.readValue(element.toString(), type);
         } catch (IOException e) {
             throw new HugeException(String.format(
-                      "Failed to deserialize %s", content), e);
+                    "Failed to deserialize %s", content), e);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:566]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.Map;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.baidu.hugegraph.api.BaseApiTest;
+import com.google.common.collect.ImmutableMap;
+
+
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java:11]: one empty line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java [hunk 27]
```diff
@@ -0,0 +1,37 @@
+package com.baidu.hugegraph.api.traversers;
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/AllShortestPathsAPITest.java:1]: add license header

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java [hunk 28]
```diff
@@ -0,0 +1,47 @@
+package com.baidu.hugegraph.api.traversers;
+
+import static com.baidu.hugegraph.testutil.Assert.assertEquals;
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/CountAPITest.java:3]: don' import static as possible

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesAPITest.java [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+        BaseApiTest.initEdge();
+    }
+
+    @Test
+    public void testList() {
+        Map<String, String> name2Ids = listAllVertexName2Ids();
+        final String edgeGetPath = "graphs/hugegraph/graph/edges";
+        String vadasId = name2Ids.get("vadas");
+        Response r = client().get(edgeGetPath,
+                                  ImmutableMap.of("vertex_id",
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesAPITest.java:41]: define var params first

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesApiTest.java [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+import com.baidu.hugegraph.api.BaseApiTest;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+
+
+public class EdgesApiTest extends BaseApiTest {
+
+    final static String EDGE_PATH = "graphs/hugegraph/traversers/edges";
+    final static String SHARES_PATH = "graphs/hugegraph/traversers"
+                                      + "/edges/shards";
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesApiTest.java:23]: move "+" to previous line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesAPITest.java [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+import org.junit.Test;
+
+import com.baidu.hugegraph.api.BaseApiTest;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+
+
+public class EdgesApiTest extends BaseApiTest {
+
+    final static String EDGE_PATH = "graphs/hugegraph/traversers/edges";
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/EdgesAPITest.java:21]: set public, can define base class TraverserApiTest

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/TraversersApiTestSuite.java [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+    RingsApiTest.class,
+    SameNeighborsApiTest.class,
+    ShortestPathApiTest.class,
+    SingleSourceShortestPathApiTest.class,
+    TemplatePathsApiTest.class
+})
+public class TraversersApiTestSuite {
+
+    @BeforeClass
+    public static void initEnv() {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/TraversersApiTestSuite.java:34]: delete this method and add TraversersApiTestSuite into ApiTestSuite

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/KneighborApiTest.java [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    public void testPost() {
+        Map<String, String> name2Ids = listAllVertexName2Ids();
+        String markoId = name2Ids.get("marko");
+        String reqBody = String.format("{ "
+                                       + "\"source\": \"%s\", "
+                                       + "\"step\": { "
+                                       + "  \"direction\": \"BOTH\", "
+                                       +
+                                       "  \"labels\": [\"knows\", " +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/KneighborApiTest.java:68]: keep same style

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/KoutApiTest.java [hunk 34]
```diff
@@ -0,0 +1,93 @@
...
+    @Test
+    public void testPost() {
+        Map<String, String> name2Ids = listAllVertexName2Ids();
+        String markoId = name2Ids.get("marko");
+        String reqBody = String.format("{ "
+                                       + "\"source\": \"%s\", "
+                                       + "\"step\": { "
+                                       + "  \"direction\": \"BOTH\", "
+                                       +
+                                       "  \"labels\": [\"knows\", " +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/KoutApiTest.java:73]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/PathsApiTest.java [hunk 35]
```diff
@@ -0,0 +1,90 @@
...
+                                       + "    \"weight\": \"P.gt(0.01)\"}}, "
+                                       + "\"max_depth\": 10, "
+                                       + "\"capacity\": 100000000, "
+                                       + "\"limit\": 10000000, "
+                                       + "\"with_vertex\": false}",
+                                       markoId, joshId);
+        Response r = client().post(path, reqBody);
+        String respBody = assertResponseStatus(200, r);
+        List<Map<String, Object>> paths = assertJsonContains(respBody, "paths");
+        Assert.assertFalse(paths.isEmpty());
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/PathsApiTest.java:88]: Assert.assertEquals(xx, paths.size());

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/PersonalRankApiTest.java [hunk 36]
```diff
@@ -0,0 +1,60 @@
...
+        String reqBody = String.format("{"
+                                       + "\"source\":\"%s\","
+                                       + "\"max_depth\":\"%s\","
+                                       + "\"label\":\"%s\","
+                                       + "\"alpha\":\"%s\"}",
+                                       markoId, 3, "created", 1);
+        Response r = client().post(path, reqBody);
+        String respBody = assertResponseStatus(200, r);
+        Map<String, Object> entity = parseMap(respBody);
+        Assert.assertNotNull(entity);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/PersonalRankApiTest.java:58]: assert more results content

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/RingsApiTest.java [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+        Map<String, String> name2Ids = listAllVertexName2Ids();
+        String markoId = name2Ids.get("marko");
+        Map<String, Object> params = ImmutableMap.of("source",
+                                                     id2Json(markoId),
+                                                     "max_depth", 10);
+        Response r = client().get(path, params);
+        String respJson = assertResponseStatus(200, r);
+        List<Map<String, List<String>>> rings = assertJsonContains(respJson,
+                                                                   "rings");
+        Assert.assertFalse(rings.isEmpty());
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/RingsApiTest.java:58]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/ShortestPathApiTest.java [hunk 38]
```diff
@@ -0,0 +1,25 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.baidu.hugegraph.api.traversers;
+
+import com.baidu.hugegraph.api.BaseApiTest;
+
+public class ShortestPathApiTest extends BaseApiTest {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/ShortestPathApiTest.java:24]: no test case

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java [hunk 39]
```diff
@@ -0,0 +1,25 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.baidu.hugegraph.api.traversers;
+
+import com.baidu.hugegraph.api.BaseApiTest;
+
+public class SingleSourceShortestPathApiTest extends BaseApiTest {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java:24]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/TemplatePathsApiTest.java [hunk 40]
```diff
@@ -0,0 +1,25 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.baidu.hugegraph.api.traversers;
+
+import com.baidu.hugegraph.api.BaseApiTest;
+
+public class TemplatePathsApiTest extends BaseApiTest {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/TemplatePathsApiTest.java:24]: ditto

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/KoutApiTest.java [hunk 41]
```diff
@@ -51,13 +51,25 @@ public void testGet() {
...
+        String rippleId = name2Ids.get("ripple");
+        // Test for nearest is true
         Response r = client().get(path,
                                   ImmutableMap.of("source", id2Json(markoId),
                                                   "max_depth", 2));
         String respBody = assertResponseStatus(200, r);
         List<String> vertices = assertJsonContains(respBody, "vertices");
         Assert.assertEquals(1, vertices.size());
+        Assert.assertTrue(vertices.contains(joshId));
+        // Test for nearest is false
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/KoutApiTest.java:63]: nearest=false

File: hugegraph-dist/pom.xml [hunk 42]
```diff
@@ -23,6 +23,12 @@
             <groupId>com.baidu.hugegraph</groupId>
             <artifactId>hugegraph-core</artifactId>
             <version>${project.version}</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.google.protobuf</groupId>
+                    <artifactId>protobuf-java</artifactId>
```
Comment (javeme) [inline hugegraph-dist/pom.xml:29]: which package is it from

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathApiTest.java [hunk 43]
```diff
@@ -0,0 +1,70 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import com.baidu.hugegraph.api.BaseApiTest;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+public class WeightedShortestPathApiTest extends BaseApiTest {
+
+    final static String path = "graphs/hugegraph/traversers/" +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathApiTest.java:37]: define var TRAVERSERS_API = "graphs/hugegraph/traversers/"

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java [hunk 44]
```diff
@@ -0,0 +1,59 @@
...
+    @Test
+    public void testGet() {
+        Map<String, String> name2Ids = listAllVertexName2Ids();
+        String markoId = name2Ids.get("marko");
+        Response r = client().get(path, ImmutableMap.of("source",
+                                                        id2Json(markoId),
+                                                        "with_vertex",
+                                                        true));
+        String respBody = assertResponseStatus(200, r);
+        Map<String, Map> paths = assertJsonContains(respBody,"paths");
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java:56]: expect space

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java [hunk 45]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Test
+    public void testGet() {
+        Map<String, String> name2Ids = listAllVertexName2Ids();
+        String markoId = name2Ids.get("marko");
+        Response r = client().get(path, ImmutableMap.of("source",
+                                                        id2Json(markoId),
+                                                        "with_vertex",
+                                                        true));
+        String respBody = assertResponseStatus(200, r);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java:55]: String content

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java [hunk 46]
```diff
@@ -0,0 +1,59 @@
...
+        Map<String, Map> paths = assertJsonContains(respBody,"paths");
+        Assert.assertFalse(paths.isEmpty());
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/SingleSourceShortestPathApiTest.java:57]: assert paths.size()

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathApiTest.java [hunk 47]
```diff
@@ -0,0 +1,70 @@
...
+        String joshId = name2Ids.get("josh");
+        String peterId = name2Ids.get("peter");
+        String rippleId = name2Ids.get("ripple");
+        Response r = client().get(path, ImmutableMap.of("source",
+                                                        id2Json(markoId),
+                                                        "target",
+                                                        id2Json(joshId),
+                                                        "weight", "weight",
+                                                        "with_vertex", true));
+        String respBody = assertResponseStatus(200, r);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathApiTest.java:62]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathApiTest.java [hunk 48]
```diff
@@ -0,0 +1,70 @@
...
+        Map<String, Map> paths = assertJsonContains(respBody,"path");
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathApiTest.java:63]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/ShortestPathApiTest.java [hunk 49]
```diff
@@ -0,0 +1,62 @@
...
+        String joshId = name2Ids.get("vadas");
+        Response r = client().get(path, ImmutableMap.of("source",
+                                                        id2Json(markoId),
+                                                        "target",
+                                                        id2Json(joshId),
+                                                        "max_depth",
+                                                        100));
+        String respBody = assertResponseStatus(200, r);
+        List<String> paths = assertJsonContains(respBody, "path");
+        Assert.assertFalse(paths.isEmpty());
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/traversers/ShortestPathApiTest.java:60]: can assert the size and elements?

**Review by javeme — APPROVED** [review]

**Review by zhoney — APPROVED** [review]

**Pull request merged**